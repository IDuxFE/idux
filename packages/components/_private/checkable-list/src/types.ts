/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
