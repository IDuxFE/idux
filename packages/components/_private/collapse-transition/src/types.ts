/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
