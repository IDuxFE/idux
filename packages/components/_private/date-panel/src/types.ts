/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
