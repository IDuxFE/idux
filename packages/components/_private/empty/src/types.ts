/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
