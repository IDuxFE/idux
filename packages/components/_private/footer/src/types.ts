/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
