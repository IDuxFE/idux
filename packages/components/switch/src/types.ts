import type { DefineComponent, HTMLAttributes } from 'vue'
import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'

import { IxPropTypes } from '@idux/cdk/utils'

export const switchProps = {
  checked: IxPropTypes.bool.def(false),
  disabled: IxPropTypes.bool.def(false),
  checkedChildren: IxPropTypes.string.def(''),
  unCheckedChildren: IxPropTypes.string.def(''),
  size: IxPropTypes.oneOf(['medium', 'small'] as const).def('medium'),
  loading: IxPropTypes.bool.def(false),
}

export type SwitchProps = IxInnerPropTypes<typeof switchProps>
export type SwitchPublicProps = IxPublicPropTypes<typeof switchProps>
export interface SwitchBindings {
  focus: (options?: FocusOptions) => void
  blur: () => void
}
export type SwitchComponent = DefineComponent<HTMLAttributes & typeof switchProps, SwitchBindings>
export type SwitchInstance = InstanceType<DefineComponent<SwitchProps, SwitchBindings>>
