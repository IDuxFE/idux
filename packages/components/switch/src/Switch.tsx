/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, onMounted } from 'vue'

import { callEmit } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { FORM_TOKEN } from '@idux/components/form'
import { IxIcon } from '@idux/components/icon'
import { useFormAccessor, useFormElement } from '@idux/components/utils'

import { switchProps } from './types'

export default defineComponent({
  name: 'IxSwitch',
  props: switchProps,
  setup(props, { expose, slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-switch`)
    const { elementRef, focus, blur } = useFormElement()
    const formContext = inject(FORM_TOKEN, null)

    expose({ focus, blur })

    const accessor = useFormAccessor<boolean>('checked')

    const isChecked = computed(() => accessor.valueRef.value)
    const isDisabled = computed(() => props.disabled ?? accessor.disabled.value)

    const handleClick = () => {
      if (isDisabled.value || props.loading) {
        return
      }
      callEmit(props.onChange, !isChecked.value)
      accessor.setValue(!isChecked.value)
    }

    const handleFocus = (evt: FocusEvent) => {
      callEmit(props.onFocus, evt)
    }

    const handleBlur = (evt: FocusEvent) => {
      callEmit(props.onBlur, evt)
      accessor.markAsBlurred()
    }

    const size = computed(() => props.size ?? formContext?.size.value)
    const classes = computed(() => {
      const { loading } = props
      const prefixCls = mergedPrefixCls.value
      return {
        [prefixCls]: true,
        [`${prefixCls}-checked`]: isChecked.value,
        [`${prefixCls}-disabled`]: isDisabled.value,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-${size.value}`]: true,
      }
    })

    onMounted(() => props.autofocus && focus())

    return () => {
      const checked = isChecked.value
      const label = slots.label?.({ checked }) ?? props.labels[checked ? 0 : 1]
      const prefixCls = mergedPrefixCls.value
      return (
        <button
          ref={elementRef}
          type="button"
          class={classes.value}
          onClick={handleClick}
          onFocus={handleFocus}
          onBlur={handleBlur}
        >
          {props.loading && (
            <span class={`${prefixCls}-loading-icon`}>
              <IxIcon name="loading" />
            </span>
          )}
          <span class={`${prefixCls}-label`}>{label}</span>
        </button>
      )
    }
  },
})
