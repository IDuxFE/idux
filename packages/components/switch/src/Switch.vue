<template>
  <button
    ref="switchRef"
    class="ix-switch"
    :class="[
      { 'ix-switch-checked': stateChecked },
      { 'ix-switch-disabled': disabled || loading },
      { 'ix-switch-small': isSmallSize },
    ]"
    type="button"
    @mouseup="handleMouseup"
    @click="handleClick"
  >
    <div v-if="loading" class="ix-switch-loading">
      <ix-icon name="loading" />
    </div>
    <div class="ix-switch-inner">
      <template v-if="stateChecked">
        <slot name="checkedChildren">{{ checkedChildren }}</slot>
      </template>
      <template v-else>
        <slot name="unCheckedChildren">{{ unCheckedChildren }}</slot>
      </template>
    </div>
  </button>
</template>
<script lang="ts">
import { computed, defineComponent, ref, watch } from 'vue'
import { SwitchProps } from './types'
import { PropTypes } from '@idux/cdk/utils'
import { IxIcon } from '@idux/components/icon'

export default defineComponent({
  name: 'IxSwitch',
  components: { IxIcon },
  inheritAttrs: false,
  props: {
    checked: PropTypes.bool.def(false),
    disabled: PropTypes.bool.def(false),
    checkedChildren: PropTypes.string.def(''),
    unCheckedChildren: PropTypes.string.def(''),
    size: PropTypes.oneOf(['medium', 'small'] as const).def('medium'),
    loading: PropTypes.bool.def(false),
  },
  emits: ['change', 'update:checked'],
  setup(props: SwitchProps, { emit }) {
    const stateChecked = ref(props.checked)
    const switchRef = ref<HTMLButtonElement | null>(null)

    const isSmallSize = computed(() => props.size === 'small')

    watch(
      () => props.checked,
      curChecked => {
        stateChecked.value = curChecked
      },
    )

    const handleClick = () => {
      if (props.disabled || props.loading) {
        return
      }
      const checked = !stateChecked.value
      stateChecked.value = checked
      emit('update:checked', checked)
      emit('change', checked)
    }

    const handleMouseup = () => {
      blur()
    }

    const focus = () => {
      switchRef.value?.focus()
    }

    const blur = () => {
      switchRef.value?.blur()
    }

    return {
      stateChecked,
      handleClick,
      isSmallSize,
      handleMouseup,
      switchRef,
      focus,
      blur,
    }
  },
})
</script>
