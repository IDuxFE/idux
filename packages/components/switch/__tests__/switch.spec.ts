import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import IxSwitch from '../src/Switch'
import { SwitchProps } from '../src/types'

describe('Switch', () => {
  const SwitchMount = (options?: MountingOptions<Partial<SwitchProps>>) => mount(IxSwitch, { ...options })

  renderWork(IxSwitch)

  test('v-model:checked work', async () => {
    const onUpdateChecked = jest.fn()
    const wrapper = SwitchMount({ props: { checked: true, 'onUpdate:checked': onUpdateChecked } })

    expect(wrapper.classes()).toContain('ix-switch-checked')

    await wrapper.setProps({ checked: false })

    expect(wrapper.classes()).not.toContain('ix-switch-checked')

    await wrapper.trigger('click')

    expect(onUpdateChecked).toBeCalledWith(true)
  })

  test('labels work', async () => {
    const wrapper = SwitchMount({
      props: { checked: true, labels: ['open', 'close'] },
    })

    expect(wrapper.find('.ix-switch-label').text()).toEqual('open')

    await wrapper.setProps({ checked: false })

    expect(wrapper.find('.ix-switch-label').text()).toEqual('close')
  })

  test('label slot work', async () => {
    const wrapper = SwitchMount({
      props: { checked: true, labels: ['open', 'close'] },
      slots: {
        label: ({ checked }: { checked: boolean }) => (checked ? 'open slot' : 'close slot'),
      },
    })

    expect(wrapper.find('.ix-switch-label').text()).toEqual('open slot')

    await wrapper.setProps({ checked: false })

    expect(wrapper.find('.ix-switch-label').text()).toEqual('close slot')
  })

  test('disabled work', async () => {
    const wrapper = SwitchMount({
      props: { disabled: true },
    })

    await wrapper.trigger('click')

    expect(wrapper.classes()).toContain('ix-switch-disabled')

    await wrapper.setProps({ disabled: false })

    await wrapper.trigger('click')

    expect(wrapper.classes()).not.toContain('ix-switch-disabled')
  })

  test('loading work', async () => {
    const wrapper = SwitchMount({ props: { loading: true } })

    expect(wrapper.find('.ix-switch-loading-icon').exists()).toBeTruthy()
    expect(wrapper.classes()).toContain('ix-switch-loading')

    await wrapper.trigger('click')

    await wrapper.setProps({ loading: false })

    expect(wrapper.find('.ix-switch-loading-icon').exists()).toBeFalsy()
    expect(wrapper.classes()).not.toContain('ix-switch-loading')
  })

  test('size work', async () => {
    const wrapper = SwitchMount({ props: { size: 'sm' } })

    expect(wrapper.classes()).toContain('ix-switch-sm')

    await wrapper.setProps({ size: undefined })

    expect(wrapper.classes()).toContain('ix-switch-md')
  })
})
