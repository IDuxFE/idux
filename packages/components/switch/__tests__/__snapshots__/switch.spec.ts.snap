// Vitest Snapshot v1

exports[`Switch > render work 1`] = `
"<button type=\\"button\\" class=\\"ix-switch ix-switch-md\\">
  <!----><span class=\\"ix-switch-label\\"><!----></span>
</button>"
`;
